### PR TITLE
Add E2E Factories

### PR DESCRIPTION
**Pull request by ObliviousHarmony**
### All Submissions:

* [ ] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [ ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

There is currently no easy way to generate database models for our E2E tests. At the moment we're required to run everything against the `wp-admin` UI, which is not going to be the fastest way to create test data. One other option would be a database or CSV import, but I feel like those options lack the control and accessibility of a factory pattern tied to the REST API.

I propose that we create a series of factories for the various data types, as well as some functions that map to API calls to set common settings use-cases. I would also like to touch on the two larger elephants in this PR, the adoption of TypeScript, as well as the [Fishery](https://github.com/thoughtbot/fishery) package from Thoughtbot.

With respect to TypeScript, I am a huge proponent of type systems. I believe they eliminate an entire class of bugs almost effortlessly, as well as provide a lot of interesting information for optimization to compilers. While introducing TypeScript may potentially deter contributions in a larger project, I feel that this package's surface area is small enough that it won't be problematic. For those that use TypeScript, this package will provide a great deal of safety, and for those that don't, it will be as if it doesn't exist.

As for Fishery, as part of my initial research I looked at basically every JavaScript factory library. When I found Fishery I fell in love, as I felt like it was exactly what I would have written if I had attempted to roll my own solution. While it has not yet reached 1.0, I think that the desire to keep the API in-line with Factory_Bot will prevent it from being too unstable externally, so I don't have too much concern about that. **Currently the PR has included the library internally, as the code support `async` creation has not yet been merged**.

**WIP Notes:**
- Currently the E2E suite is failing because of our dependency on `@wordpress/api-fetch`. There is no `window.fetch` to utilize apparently, but I wonder why given that it should be running in Puppeteer?

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

### CODE REVIEW

File: tests/e2e-tests/specs/front-end/front-end-cart.test.js
Comment (senadir) [inline tests/e2e-tests/specs/front-end/front-end-cart.test.js:21]: depends on your setup, but it's much preferable to run a fixture import in the global setup phase so that you only import once

File: tests/e2e/factories/package.json
Comment (senadir) [inline tests/e2e/factories/package.json:25]: I share Darren concern, why are we including this manually?

File: tests/e2e/factories/src/factories/product.ts
Comment (senadir) [inline tests/e2e/factories/src/factories/product.ts:19]: As Darren noted in slack, this won't work because Jest is running the code not puppeteer, you don't run code in puppeteer you control the browser via a set of utilities.

I suggest we use https://github.com/node-fetch/node-fetch to keep the same API and spec.

File: tests/e2e/factories/src/models/model.ts
Comment (senadir) [inline tests/e2e/factories/src/models/model.ts:20]: should we throw errors if the module is already created? why not skip or update silently?

**Review by senadir — CHANGES_REQUESTED** [review]
This is looking like a good start so far, I left some comments.

File: tests/e2e/factories/package.json
Comment (ObliviousHarmony) [reply tests/e2e/factories/package.json:25]: We're currently blocked on including via NPM by [this issue](https://github.com/thoughtbot/fishery/issues/16). The version that we're including here comes from the `async` branch, as I didn't want to be stalled over the weekend in working on this.

File: tests/e2e/factories/src/factories/product.ts
Comment (ObliviousHarmony) [reply tests/e2e/factories/src/factories/product.ts:19]: I agree, I was looking at that library last week but there was a misunderstanding around the availability of browser globals in the suite. Thank you very much for confirming that this package would be a good choice!

File: tests/e2e/factories/src/models/model.ts
Comment (ObliviousHarmony) [reply tests/e2e/factories/src/models/model.ts:20]: I'm generally a fan of throwing errors as explicitly and as early as possible. I find that programming defensively prevents people from using things in a way that we didn't intend (because they almost always do!)

File: tests/e2e-tests/specs/front-end/front-end-cart.test.js
Comment (ObliviousHarmony) [reply tests/e2e-tests/specs/front-end/front-end-cart.test.js:21]: I usually prefer to keep test data generation as close to the tests as possible in order to allow for parallelization. That said, this was just intended to be an example usage of the package, as opposed to a solid decision about how we're going to consume it in Core.

**Comment by senadir** [comment]
What is currently left before we can merge this?

**Comment by ObliviousHarmony** [comment]
> What is currently left before we can merge this?

- We're currently blocked by the `fishery` package. I received a reply from the author and have submitted some code. The resolution of this issue may result in changes to the factory API, so I'm waiting before pushing this forward onto anyone. This doesn't stop the work that can be done in the PR of course.
- `node-sass` needs to be added to polyfill `window.fetch`.
- The `Product` model + factory need to be fully fleshed out and tested.
- Factories need to be created for all other core data types.

I think that would be a good start, after which we can look at convenience methods for settings changes and things like that, as well as possible REST API extensions where needed.

**Comment by ObliviousHarmony** [comment]
I've had some time the last two weekends to work on furthering this, and I've got an update on some decisions that have been made!

- We're going to need to switch `@wordpress/api-fetch` out for `axios`. Due to the global nature of `apiFetch` as well as the need to polyfill `window.fetch`, it isn't looking like it is a very good fit.
- A [PR in the fishery package](https://github.com/thoughtbot/fishery/pull/20) will allow us to drop the included package in favor of the NPM one. It supports extending factories which will allow us to build our REST API functionality on top of the package without needing to abstract away its API.

**Comment by ObliviousHarmony** [comment]
Closed in favor of #26948

**Pull request closed without merging**
